### PR TITLE
Correctly type `GraphQLHookOptions`

### PR DESCRIPTION
**Pull request by corbt**
Currently, completing [tutorial 2](https://learn.redwoodjs.com/docs/tutorial2/creating-a-comment-form#graphql-query-caching) in TS throws a type error because the `onCompleted` and `refetchQueries` keys aren't permitted in `GraphQLHookOptions`.

I've also added a catch-all `[key: string]: any`. Since `GraphQLHooksOptions` just passes its options into the underlying implementation, it's possible that the user's implementation will accept/expect more keys, and those shouldn't throw an error if provided.

cc @peterp

### CODE REVIEW

**Comment by Tobbe** [comment]
/cc @dac09 
I know you've been doing some related TS work lately, so tagging you

**Comment by dac09** [comment]
Thanks for raising this! I'll be looking at this soon :), will ping you here if I need some input!

**Comment by dac09** [comment]
This looks ok to me, certainly an improvement on what we currently have! 

I haven't looked too deeply into this, I wonder if we can dynamically get the type from `useQuery` and `useMutation` forward on? I'm also not sure why options are currently shared between `useQuery` and `useMutation` since their options are different. 

Any insight into this @corbt ?

**Review by dac09 — APPROVED** [review]

File: packages/web/src/components/GraphQLHooksProvider.tsx [hunk 1]
```diff
@@ -2,6 +2,9 @@ import type { DocumentNode } from 'graphql'
 
 export interface GraphQLHookOptions {
   variables?: Record<string, any>
+  refetchQueries?: { query: DocumentNode }[]
+  onCompleted?: () => any
```
Comment (dac09) [inline packages/web/src/components/GraphQLHooksProvider.tsx:6]:
```suggestion
  onCompleted?: (data: any) => void
```

Comment (dac09) [reply packages/web/src/components/GraphQLHooksProvider.tsx:6]: I know this is not perfect at the moment, I'll have to take a deeper look into typing this more completely

**Comment by corbt** [comment]
> I haven't looked too deeply into this, I wonder if we can dynamically get the type from `useQuery` and `useMutation` forward on? I'm also not sure why options are currently shared between `useQuery` and `useMutation` since their options are different.

I agree, fundamentally `useQuery` and `useMutation` have different types and their type signatures shouldn't be unified. This was just the quickest fix to get rid of the red squigglies when going through the tutorial!

**Comment by thedavidprice** [comment]
@dac09 did you want this in as is? Or waiting to loop in Peter?

**Comment by peterp** [comment]
I think we should make sure that these types can be enhanced, so that the react query library can also work with this.

File: packages/web/src/components/GraphQLHooksProvider.tsx [hunk 2]
```diff
@@ -2,6 +2,9 @@ import type { DocumentNode } from 'graphql'
 
 export interface GraphQLHookOptions {
   variables?: Record<string, any>
+  refetchQueries?: { query: DocumentNode; variables?: Record<string, any> }[]
+  onCompleted?: () => any
```
Comment (dac09) [inline packages/web/src/components/GraphQLHooksProvider.tsx:6]: The suggested change got reverted @corbt after the force push :).

Comment (corbt) [reply packages/web/src/components/GraphQLHooksProvider.tsx]: Drat, that's what I get for trying to solve problems late at night 🤣 . Re-added.

**Pull request merged**